src/cpp/session/modules/rmarkdown/NotebookAlternateEngines.cpp

```diff
@@ -0,0 +1,404 @@
+/*
+ * NotebookAlternateEngines.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+
+#include "SessionRmdNotebook.hpp"
+#include "SessionExecuteChunkOperation.hpp"
+#include "NotebookCache.hpp"
+#include "NotebookAlternateEngines.hpp"
+
+#include <boost/algorithm/string.hpp>
+
+#include <core/StringUtils.hpp>
+#include <core/Algorithm.hpp>
+#include <core/Exec.hpp>
+#include <core/json/JsonRpc.hpp>
+
+#include <r/RExec.hpp>
+
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+namespace {
+
+class ChunkExecCompletedScope
```

Comment:
I think this should inherit from boost::noncopyable (since the destructor is non-trivial)